### PR TITLE
Fix sds bootstrap configs for Istio control plane

### PR DESCRIPTION
**Pull request by pitlv2109**
* Fix SDS bootstrap config for Pilot, Galley, and Mixer. Prior to this, even if sds is used, aforementioned components still used file mount certs. This will break Istio if the control plane is set to use an SDS-based CA (pods like ingress gateway will fail to connect to Pilot, etc. due to having different root certs).

* Configure the Mixer checker server to use sidecar for MCP connections to align with the rest of Istio control plane components.

### CODE REVIEW

**Comment by pitlv2109** [comment]
Please cc relevant people from Pilot, Galley, and Mixer. I tested it locally with security features and everything looks fine.

**Review by myidpt — APPROVED** [review]

File: istio-policy/templates/deployment.yaml
Comment (JimmyCYJ) [inline istio-policy/templates/deployment.yaml]: Is this change consistent with the old installer? I am not clear why this is needed.

**Review by JimmyCYJ — APPROVED** [review]

File: istio-policy/templates/deployment.yaml
Comment (pitlv2109) [reply istio-policy/templates/deployment.yaml]: This is for the policy mixer container to talk to its sidecar, similar to https://github.com/istio/installer/pull/108. In the old installer, Galley doesn't have a sidecar and apparently policy-mixer bypasses its sidecar and talks to Galley directly.

**Review by sdake — CHANGES_REQUESTED** [review]
this is a pretty big change for a bunch of yaml :).  What was the source of these changes?  Manually generated, or came from some other repository?

Cheers
-steve

File: istio-control/istio-config/templates/deployment.yaml
Comment (sdake) [inline istio-control/istio-config/templates/deployment.yaml]: why this expirationSeconds magic number?

File: istio-control/istio-config/templates/configmap-envoy.yaml
Comment (sdake) [inline istio-control/istio-config/templates/configmap-envoy.yaml]: `h2` is some type of alpn protocol?

File: istio-control/istio-config/templates/deployment.yaml
Comment (sdake) [inline istio-control/istio-config/templates/deployment.yaml]: will this not collide with K8s's `/var/run/secrets/tokens` mount?

File: istio-control/istio-config/templates/deployment.yaml
Comment (sdake) [inline istio-control/istio-config/templates/deployment.yaml]: why a hostpath?  There is no good reason to use a host path in anywhere of the manifests of Istio with the possible exception of CNI.

File: istio-control/istio-discovery/templates/configmap-envoy.yaml
Comment (sdake) [inline istio-control/istio-discovery/templates/configmap-envoy.yaml]: duplicated at 112...

File: istio-control/istio-discovery/templates/configmap-envoy.yaml
Comment (sdake) [inline istio-control/istio-discovery/templates/configmap-envoy.yaml]: no underscore in sdsstat?

**Review by sdake — CHANGES_REQUESTED** [review]
once the hostpath problem is fixed (or explained in great detail) - and the other nitpicks are addressed or confirmed as accurate, this (complex to review) PR should be ready to go.  I don't have 100% confidence this code is functional - we can always to a partial or full revert if need be.

**Comment by sdake** [comment]
I'd prefer not to cherrypick this right away (or approve the cherrypick right away) for 1.4.  This is a large change that is not suitable for a backport.  I understand the significance of missing SDS, so willing to defer to the release managers on 1.4 for this PR, however, I really think it needs some baking time in master for a few weeks before backporting given its heft.

**Comment by pitlv2109** [comment]
> this is a pretty big change for a bunch of yaml :). What was the source of these changes? Manually generated, or came from some other repository?
> 
> Cheers
> -steve

The bootstrap configurations are from the old installer (e.g https://github.com/istio/istio/tree/master/pilot/docker).

File: istio-control/istio-discovery/templates/configmap-envoy.yaml
Comment (pitlv2109) [reply istio-control/istio-discovery/templates/configmap-envoy.yaml]: This is not a duplicate. Each one for a if-else branch

File: istio-control/istio-config/templates/deployment.yaml
Comment (pitlv2109) [reply istio-control/istio-config/templates/deployment.yaml]: From the old installer and this is what we've been using. I'd rather not changing any constants in this PR.

File: global.yaml
Comment (costinm) [inline global.yaml]: Why empty ? The feature is enabled by sds.enabled, no need to have to configure the path too.

File: istio-control/istio-config/templates/configmap-envoy.yaml
Comment (costinm) [inline istio-control/istio-config/templates/configmap-envoy.yaml]: Pilot has support for K8S signed certificates - which are needed for inject, etc.
And the feedback in ToC was to remove the sidecar if I remember correctly.

**Review by costinm — COMMENTED** [review]
For 1.5 - this conflicts with the plan to use K8S-signed certs and use istiod mode ( injector, etc in pilot ). 

For 1.4 - this seems far too large change. 

For both - it's quite complicated, and we already have a lot of problems with helm template complexity.

Better approach: focus on using SDS in pilot-agent and turn on SDS-only in istiod, it doesn't require any installer change except mounting the token (which can be optional, so no options), it's all in the code.

File: istio-control/istio-config/templates/deployment.yaml
Comment (myidpt) [reply istio-control/istio-config/templates/deployment.yaml]: This is 12 hours. We made this the default JWT lifetime. I think it's a balance of making the JWT short-lived VS. not causing too much JWT rotation overhead.

File: istio-control/istio-config/templates/deployment.yaml
Comment (myidpt) [reply istio-control/istio-config/templates/deployment.yaml]: Yes we agree. The reason for doing that is in we are still using the node agent DaemonSet for SDS. We are working on the per-pod Istio Agent to replace the node agent, which will get rid of the hostPath.

File: istio-control/istio-config/templates/deployment.yaml
Comment (myidpt) [reply istio-control/istio-config/templates/deployment.yaml]: Good question. This one is to load the trustworthy JWT to the "tokens" path. The default K8s JWT is in the "kubernetes.io/serviceaccount" path. I don't see they collide since they are using different sub-directories.

**Pull request closed without merging**
